File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.cs
public class Solution {
    public int MinCut(string s) {
        if(s.Length == 1)
        {
            return 0;
        }
        int[] dp = new int[s.Length+1];
        for(int i=0; i<s.Length; i++)
        {
            dp[i] = -1;
        }
        return Helper(dp, 0, s)-1;
    }
    private int Helper(int[] dp, int i, string s)
    {
        if(i == s.Length)
        {
            return 0;
        }
        if(dp[i] != -1)
        {
            return dp[i];
        }
        string current = "";
        int min = s.Length - i;
        for(int j=i; j<s.Length; j++)
        {
            current+=s[j];
            if(isPalindrome(current))
            {
                min = Math.Min(min, 1+Helper(dp, j+1, s));
            }
        }
        dp[i] = min;
        return dp[i];
    }
    private bool isPalindrome(string s)
    {
        int i=0;
        int j=s.Length-1;
        while(i<j)
        {
            if(s[i]!=s[j])
            {
                return false;
            }
            i++;
            j--;
        }
        return true;
    }
}

File: 0132-palindrome-partitioning-ii/0132-palindrome-partitioning-ii.java
class Solution {
    public int minCut(String s) {
        int n = s.length();
        // isPalindrome snippet is re-usable and almost exact answer for longest palindromic substring
        boolean[][] isPalindrome = new boolean[n][n];
        for(int i=0; i<n; i++)
        {
            isPalindrome[i][i] = true;
            if(i < n-1 && s.charAt(i) == s.charAt(i+1))
            {
                isPalindrome[i][i+1] = true;
            }
        }
        for(int length = 3; length <= n; length++)
        {
            for(int start = 0; start <= n-length; start++)
            {
                isPalindrome[start][start+length-1] = isPalindrome[start+1][start+length-2] && s.charAt(start) == s.charAt(start+length-1);
            }
        }
        int[] cache = new int[n];
        Arrays.fill(cache, -1);
        return getAns(0, s, cache, isPalindrome) - 1;
    }
    private int getAns(int currentIndex, String s, int[] cache, boolean[][] isPalindrome)
    {
        if(currentIndex == s.length())
        {
            return 0;
        }
        if(cache[currentIndex] != -1)
        {
            return cache[currentIndex];
        }
        int ans = s.length() + 1;
        for(int i=currentIndex; i<s.length(); i++)
        {
            if(isPalindrome[currentIndex][i])
            {
                ans = Math.min(ans, 1 + getAns(i+1, s, cache, isPalindrome));
            }
        }
        return cache[currentIndex] = ans;
    }
}